core/http/client/src/main/java/org/eclipse/rdf4j/http/client/SPARQLProtocolSession.java

```diff
@@ -1055,6 +1062,21 @@ protected HttpResponse execute(HttpUriRequest method) throws IOException, RDF4JE
 		}
 	}
 
+	private boolean exceptionIs(HttpResponse response, String exceptionName) {
+		Header[] headers = response.getHeaders("X-Eclipse-RDF4J-Exception");
```

Comment:
Went with `X-Eclipse-RDF4J-Exception´as the header, and it should contain a class name.

Reply:
If we must use a custom header, this seems like a good name for it. I'd remove the `X-` prefix though. That is no longer recommend practice for custom headers. See https://stackoverflow.com/questions/3561381/custom-http-headers-naming-conventions

Reply:
Alternatively: instead of using a custom header for this (which on some servers/firewalls is known to cause problems), we could also just let the client look at the response status code, and if it's a 409 conflict, see if there is a payload with a validation report in it.

Reply:
We shouldn’t rely on having to look at the content, since that will stop us from doing a streaming based approach down the line. Custom status line was my first idea until I realized how hard it would be to do with spring.

Reply:
Fwiw I'm fine with going with the header for now but I don't see how looking at the content would stop us streaming: based on content type you can just immediately parse in the background, this is no different from how query results are processed. Also: do these reports really get so large that this is a serious consideration?

Reply:
Can you point me to where this is handled for the query results?

My fear is that someone will modify a shape and that will cause a lot of data to become invalid. I've done this myself a few times while developing validation rules. It's usually not intentional, but it would be a pity if we bring down the server because we run out of memory.

Reply:
> Can you point me to where this is handled for the query results?

Sure thing. For construct queries, for example:
- server-side streaming: `org.eclipse.rdf4j.http.server.repository.GraphQueryResultView` (the `renderInternal` method)
- client-side streaming: `SPARQLProtocolSession.getRDFBackground`.

Reply:
SPARQLProtocolSession.getRDFBackground is based on headers and not content. It gets the content type from the headers and then passes the content of the response to the parser. It never checks anything from the content.

We could extend the content type, essentially like how application/json is extended to application/json+ld we could extend application/n-quads to be application/n-quads+shacl-validation-report or application/json+ld+shacl-validation-report.

I still think I prefer a custom header. I'll post this to the SHACL forum, maybe someone has come up with something already that we can consider implementing too.

Reply:
> SPARQLProtocolSession.getRDFBackground is based on headers and not content. It gets the content type from the headers and then passes the content of the response to the parser. It never checks anything from the content.

Yes, I know, that's what I meant. Sorry if this wasn't clear, my point was that based purely on the status code and the response content type you can process the response body (streaming). No need for an additional custom header.

Reply:
Andy (Jena dev) proposed the same to my email on the SHACL public mailing list. application/shacl-validation-report+n-quads. I’m starting to appreciate that. Will see what others reply before we decide. Would be nice if at least Jena and RDF4J did it the same way.

Reply:
I hadn't even thought of a custom content type - I was just thinking of using plain `text/turtle` (or `application/n-quads`). That's all you need to parse the incoming data. Figuring out that it is a SHACL validation report can be done on the parsed model.